### ISSUE TITLE
Longhorn v0.7.0, On kubernetes 1.16.4, volumes never attach

### ISSUE THREAD

**Issue by dmayle** [longhorn/longhorn#1044]
This seems to be the same as the closed issue #771, which didn't have a resolution.  I'm upgrading my test cluster from kubernetes 1.15.3 to 1.16.4, and under 1.16.4 the volumes got stuck in a detach/attach loop.

I'm attaching a support bundle with the logs

[longhorn-support-bundle_135c5a72-0ac6-4769-9137-01b9236d3b31_2020-02-05T07-47-56Z.zip](https://github.com/longhorn/longhorn/files/4158115/longhorn-support-bundle_135c5a72-0ac6-4769-9137-01b9236d3b31_2020-02-05T07-47-56Z.zip)

**Comment by dmayle** [comment]
I think the relevant bit of the log from the attacher:
```
I0205 08:09:34.737652       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:09:39.773362       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:09:44.808410       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:09:45.885272       1 controller.go:175] Started VA processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.885374       1 csi_handler.go:89] CSIHandler: processing VA "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.885403       1 csi_handler.go:116] Attaching "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.885456       1 csi_handler.go:249] Starting attach operation for "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.885787       1 csi_handler.go:209] PV finalizer is already set on "container-registry"
I0205 08:09:45.885843       1 csi_handler.go:542] Found NodeID dev04 in CSINode dev04
I0205 08:09:45.885917       1 csi_handler.go:169] VA finalizer is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.885969       1 csi_handler.go:183] NodeID annotation is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.886013       1 connection.go:180] GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0205 08:09:45.886057       1 connection.go:181] GRPC request: {"node_id":"dev04","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"diskSelector":"","nodeSelector":"","numberOfReplicas":"\u0003","staleReplicaTimeout":"\u0000"},"volume_id":"container-registry"}
I0205 08:09:45.924901       1 connection.go:183] GRPC response: {}
I0205 08:09:45.926313       1 connection.go:184] GRPC error: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:09:45.926358       1 csi_handler.go:412] Saving attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.951984       1 csi_handler.go:423] Saved attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:09:45.952114       1 csi_handler.go:99] Error processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674": failed to attach: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:09:45.953170       1 controller.go:141] Ignoring VolumeAttachment "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674" change
I0205 08:09:49.848076       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:09:54.881673       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
```

**Comment by dmayle** [comment]
Looking at another project using CSI, it looks like longhorn is mishandling the error.  This coment:
https://github.com/intel/pmem-csi/issues/103#issuecomment-446289994

Led me to the CSI spec:
https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerpublishvolume

Which says that if the volume has already been published than it should just continue successfully, ignoring the error.

**Comment by dmayle** [comment]
To get clean logs, I scaled down the deployment trying to attach, so the volume is in a detached state before this next log run:
```
I0205 08:22:47.149587       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:22:52.191089       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:22:57.225708       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
```
Now, I scale up and get the following logs:
```
leader-driver-longhorn-io
I0205 08:23:37.556272       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:23:42.593116       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:23:43.587922       1 controller.go:175] Started VA processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.589903       1 csi_handler.go:89] CSIHandler: processing VA "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.591066       1 csi_handler.go:116] Attaching "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.591786       1 csi_handler.go:249] Starting attach operation for "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.592879       1 csi_handler.go:209] PV finalizer is already set on "container-registry"
I0205 08:23:43.593735       1 csi_handler.go:542] Found NodeID dev04 in CSINode dev04
I0205 08:23:43.594562       1 csi_handler.go:177] VA finalizer added to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.595242       1 csi_handler.go:191] NodeID annotation added to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:43.653369       1 connection.go:180] GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0205 08:23:43.654396       1 connection.go:181] GRPC request: {"node_id":"dev04","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"diskSelector":"","nodeSelector":"","numberOfReplicas":"\u0003","staleReplicaTimeout":"\u0000"},"volume_id":"container-registry"}
I0205 08:23:47.629111       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:23:52.673066       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:23:57.721620       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
I0205 08:23:58.653746       1 connection.go:183] GRPC response: {}
I0205 08:23:58.657399       1 connection.go:184] GRPC error: rpc error: code = DeadlineExceeded desc = context deadline exceeded
I0205 08:23:58.658201       1 csi_handler.go:412] Saving attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.683277       1 csi_handler.go:423] Saved attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.684089       1 csi_handler.go:99] Error processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674": failed to attach: rpc error: code = DeadlineExceeded desc = context deadline exceeded
I0205 08:23:58.684684       1 controller.go:175] Started VA processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.685090       1 csi_handler.go:89] CSIHandler: processing VA "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.685532       1 csi_handler.go:116] Attaching "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.685946       1 csi_handler.go:249] Starting attach operation for "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.686601       1 csi_handler.go:209] PV finalizer is already set on "container-registry"
I0205 08:23:58.687106       1 csi_handler.go:542] Found NodeID dev04 in CSINode dev04
I0205 08:23:58.687555       1 csi_handler.go:169] VA finalizer is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.687998       1 csi_handler.go:183] NodeID annotation is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.689544       1 connection.go:180] GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0205 08:23:58.689263       1 controller.go:141] Ignoring VolumeAttachment "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674" change
I0205 08:23:58.690268       1 connection.go:181] GRPC request: {"node_id":"dev04","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"diskSelector":"","nodeSelector":"","numberOfReplicas":"\u0003","staleReplicaTimeout":"\u0000"},"volume_id":"container-registry"}
I0205 08:23:58.711942       1 connection.go:183] GRPC response: {}
I0205 08:23:58.714483       1 connection.go:184] GRPC error: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:23:58.714970       1 csi_handler.go:412] Saving attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.732964       1 controller.go:141] Ignoring VolumeAttachment "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674" change
I0205 08:23:58.735538       1 csi_handler.go:423] Saved attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:58.736496       1 csi_handler.go:99] Error processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674": failed to attach: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:23:59.684878       1 controller.go:175] Started VA processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.684981       1 csi_handler.go:89] CSIHandler: processing VA "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.685010       1 csi_handler.go:116] Attaching "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.685061       1 csi_handler.go:249] Starting attach operation for "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.685368       1 csi_handler.go:209] PV finalizer is already set on "container-registry"
I0205 08:23:59.685445       1 csi_handler.go:542] Found NodeID dev04 in CSINode dev04
I0205 08:23:59.685518       1 csi_handler.go:169] VA finalizer is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.685567       1 csi_handler.go:183] NodeID annotation is already set on "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.685611       1 connection.go:180] GRPC call: /csi.v1.Controller/ControllerPublishVolume
I0205 08:23:59.685630       1 connection.go:181] GRPC request: {"node_id":"dev04","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"diskSelector":"","nodeSelector":"","numberOfReplicas":"\u0003","staleReplicaTimeout":"\u0000"},"volume_id":"container-registry"}
I0205 08:23:59.797008       1 connection.go:183] GRPC response: {}
I0205 08:23:59.798577       1 connection.go:184] GRPC error: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:23:59.798624       1 csi_handler.go:412] Saving attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.819903       1 controller.go:141] Ignoring VolumeAttachment "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674" change
I0205 08:23:59.825779       1 csi_handler.go:423] Saved attach error to "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674"
I0205 08:23:59.826188       1 csi_handler.go:99] Error processing "csi-2c77e0e26cb5feca82ac601c7406a364e9ed0c975453e24fbd12e165411a0674": failed to attach: rpc error: code = Aborted desc = The volume container-registry is attaching
I0205 08:24:02.779071       1 leaderelection.go:258] successfully renewed lease longhorn-system/external-attacher-leader-driver-longhorn-io
```

**Comment by dmayle** [comment]
Yep, that's the exact same issue.  ControllerPublishVolume times out, and then longhorn is no longer able to detect that a previous call succeeded even if it times out.  I'm going to try to look at the following things:
1. (Fix) increase timeout
2. (Fix) gracefully handle the timed out exception
3. (Workaround) figure out why ControllerPublishVolume takes so long

**Comment by dmayle** [comment]
I built a 1.16.4 cluster from scratch and was able to get longhorn attaching volumes, so it looks like this bug only happens with ControllerPublishVolume times out.

What I think the cause of my timeouts is that during deployment of the cluster, I installed minio without generating a required certificate so that I had four instances attaching and detaching (since their mounted secret wasn't available) continuously for two weeks or so (I built the cluster than got the flu and came back to it after I was better).

My workaround for the moment is to not let it get in the state where there are huge timeouts, but the underlying cause is still there, and it is a bug.  longhorn needs to be able to detect when resuming from a timed out ControllerPublishVolume if the call actually succeeded.

**Comment by yasker** [comment]
@dmayle If the volume is already attaching, Longhorn currently reports error since it's not ready yet; but if it's already attached, Longhorn will not try to attach it again.

Can you try to create a volume and attach it to the UI? The attach should take about 5 seconds maximum, otherwise something might be wrong.

**Comment by dmayle** [comment]
@yasker I've tried creating a new volume and attaching, and I've got the same problem.  While performing a cluster upgrade, longhorn gets into a state where ControllerPublishVolume times out, and then this attach problem comes up.

It's the exact same problem as linked above in the pmem-csi github, which mentions the exact same problem occurring in ceph caused by a failure to implement this part of the CSI spec.

**Comment by dmayle** [comment]
I'm still looking for workarounds to prevent the ControllerPublishVolume from timing out becuase I have a production cluster that I can't upgrade because of this bug.

**Comment by yasker** [comment]
@dmayle As you can see in https://github.com/longhorn/longhorn-manager/blob/96edc510d24a65a6ebb3dbe39480913bf3c61cd4/csi/controller_server.go#L209 , if the volume is already attached, we will not return error for CSI call. So I think the problem here might be different.

The operation through UI (e.g. create volume, attach volume) is not going through the CSI interface. Can you show us the longhorn-manager log? There should be some clues on why the attaching get stuck there.

**Comment by dmayle** [comment]
I've traced through my logs, and I finally figured out that iscsid was no longer running on the node in question.  I found this interesting, because I'd never activated the iscsid service, not realizing that it was necessary  (I added the iscsi tools, just didn't know I needed to activate the service).  I think iscsid will start up based on some iscsiadm commands.

After activating the service on all of my worker nodes, I rebuilt the cluster from scratch and put my workload onto it.  While upgrading this time, I made one change to my operating procedure: before upgrading a node, I disabled scheduling and deleted all replicas on that node.

Having done both of these things, I am able to successfully upgrade my cluster. (Well, the three leaders are upgraded, and I'm currently on my fourth of five worker nodes in this cluster).  After it completes, I'm going to run one last cluster rebuild and upgrade.  This one will be with the iscsid change, but I won't manually drain the node replicas.  Assuming everything works correctly, I'll report back here and suggest that the requirements in README.md updated so that #3 says something like, 
```
3.  `open-iscsi` has been installed on all the nodes of the Kubernetes cluster, and `iscsid` enabled.
```

**Comment by yasker** [comment]
Hi @dmayle , we've updated the README.

I wasn't sure it's about `iscsid` because if `open-iscsi` is not installed, the manager will error out immediately due to unable to satisfy the requirement. But it seems we failed to take `open-iscsi` installed but `iscsid` not running into consideration.

**Comment by dmayle** [comment]
@yasker  in my case, I found logs saying that the mount command was failing because device /dev/longhorn/my-volume didn't exist.

I saw the mention of iscsiadm in the logs, and then looked up how to use the tool online.  When I ran sample commands to check the state of iscsi, that's when I saw it reported that iscsid wasn't running.  Starting the service immediately fixed the attach problem.

I checked my other hosts (both this cluster and the production cluster), and in all other nodes, iscsid was running, even though the service was not activated, which is what led me to this.

At this point, I'm in the same place on my final upgrade test (on worker node four out of five total) without the manual drain.  The upgrade is succeeding.

As to open-iscsi starting iscsid automatically, that is not the case on my distro, but I've added it to my list of notes for bare metal machines.
